cores/esp8266/base64.cpp

```diff
@@ -35,37 +34,43 @@ extern "C" {
  * @param length size_t
  * @return String
  */
-String base64::encode(const uint8_t * data, size_t length, bool doNewLines) {
+String base64::encode(const uint8_t * data, size_t length, bool doNewLines)
+{
+    String base64;
+
     // base64 needs more size then the source data, use cencode.h macros
-    size_t size = ((doNewLines ? base64_encode_expected_len(length)
-                               : base64_encode_expected_len_nonewlines(length)) + 1);
-    char * buffer = (char *) malloc(size);
-    if(buffer) {
+    size_t size = ((doNewLines ? base64_encode_expected_len( length )
+                    : base64_encode_expected_len_nonewlines( length )) + 1);
+
+    if (base64.reserve(size))
+    {
+
         base64_encodestate _state;
-        if(doNewLines)
+        if (doNewLines)
         {
             base64_init_encodestate(&_state);
         }
         else
         {
             base64_init_encodestate_nonewlines(&_state);
         }
-        int len = base64_encode_block((const char *) &data[0], length, &buffer[0], &_state);
-        len = base64_encode_blockend((buffer + len), &_state);
 
-        String base64 = String(buffer);
-        free(buffer);
-        return base64;
+        constexpr size_t BUFSIZE = 48;
+        char buf[BUFSIZE + 1 /* newline */ + 1 /* NUL */];
+        for (size_t len = 0; len < length; len += BUFSIZE * 3 / 4)
+        {
+            size_t blocklen = base64_encode_block((const char*) data + len,
+                                                  std::min( BUFSIZE * 3 / 4, length - len ), buf, &_state);
+            buf[blocklen] = '\0';
+            base64 += buf;
+        }
+        if (base64_encode_blockend(buf, &_state))
+            base64 += buf;
+    }
+    else
+    {
+        base64 = F("-FAIL-");
     }
-    return String("-FAIL-");
-}
 
-/**
- * convert input data to base64
- * @param text const String&
- * @return String
- */
-String base64::encode(const String& text, bool doNewLines) {
-    return base64::encode((const uint8_t *) text.c_str(), text.length(), doNewLines);
+    return base64;
```

Comment:
@devyte, the C++ guru, is this the right way to minimize the copying of return values?  Copy elimination or whatever it's called?  We want to hand over the pre-existing String object (created in this method's stack) instead of doing a 2nd malloc/copy.

Reply:
Single return statement, looks good for nrvo.